stacker/plan2.py

```diff
@@ -0,0 +1,297 @@
+        return Graph(steps=self.steps, dag=self.dag.filter(step_names))
+
+    def topological_sort(self):
+        nodes = self.dag.topological_sort()
+        return [self.steps[step_name] for step_name in nodes]
+
+
+class Plan(object):
+    """A collection of :class:`Step` objects to execute.
+    The :class:`Plan` helps organize the steps needed to execute a particular
+    action for a set of :class:`stacker.stack.Stack` objects. When initialized
+    with a set of steps, it will first build a Directed Acyclic Graph from the
+    steps and their dependencies.
+    Args:
+        description (str): description of the plan.
+        graph (:class:`Graph`): a graph of steps.
+    """
+
+    def __init__(self, description=None, graph=None):
+        self.id = uuid.uuid4()
+        self.description = description
+        self.graph = graph
+
+    def outline(self, level=logging.INFO, message=""):
+        """Print an outline of the actions the plan is going to take.
+        The outline will represent the rough ordering of the steps that will be
+        taken.
+        Args:
+            level (int, optional): a valid log level that should be used to log
+                the outline
+            message (str, optional): a message that will be logged to
+                the user after the outline has been logged.
+        """
+        steps = 1
+        logger.log(level, "Plan \"%s\":", self.description)
+        for step in self.steps:
+            logger.log(
+                level,
+                "  - step: %s: target: \"%s\", action: \"%s\"",
+                steps,
+                step.short_name,
+                step.fn.__name__,
+            )
+            steps += 1
+
+        if message:
+            logger.log(level, message)
+
+    def execute(self, **kwargs):
+        return self.walk(**kwargs)
+
+    def walk(self, semaphore=None):
+        """Walks each step in the underlying graph, in topological order.
+        Args:
+            step_func (func): a function that will be called with the step.
+            semaphore (threading.Semaphore, option): a semaphore object which
+                can be used to control how many steps are executed in parallel.
+                By default, there is not limit to the amount of parallelism,
+                other than what the graph topology allows.
+        """
+
+        if not semaphore:
+            semaphore = UnlimitedSemaphore()
+
+        def walk_func(step):
+            semaphore.acquire()
+            try:
+                return step.run()
+            finally:
+                semaphore.release()
+
+        return self.graph.walk(walk_func)
+
+    @property
+    def steps(self):
+        steps = self.graph.topological_sort()
+        steps.reverse()
+        return steps
+
+    @property
+    def step_names(self):
+        return [step.name for step in self.steps]
+
+    def keys(self):
+        return self.step_names
+
+
+class UnlimitedSemaphore(object):
```

Comment:
Is this just planning ahead for multi-process/threading?

Reply:
Yeah, this is leftover from multi-threading. I'll remove this since it's unnecessary right now.